File: nodejs/checkio/src/initiation/split_pairs.ts
import assert from "assert";

function splitPairs(text: string): string[] {
	return text.length >= 2
		? [text.slice(0, 2), ...splitPairs(text.slice(2))]
		: text.length === 1
			? [`${text}_`]
			: [];
}

console.log("Example:");
console.log(JSON.stringify(splitPairs("abcd")));

// These "asserts" are used for self-checking
assert.deepStrictEqual(splitPairs("abcd"), ["ab", "cd"]);
assert.deepStrictEqual(splitPairs("abc"), ["ab", "c_"]);
assert.deepStrictEqual(splitPairs("abcdf"), ["ab", "cd", "f_"]);
assert.deepStrictEqual(splitPairs("a"), ["a_"]);
assert.deepStrictEqual(splitPairs(""), []);

console.log("Coding complete? Click 'Check Solution' to earn rewards!");

File: python/checkio/initiation/split-pairs.py
from typing import Iterable


def split_pairs(text: str) -> Iterable[str]:
    if len(text) % 2 != 0:
        text += "_"
    return [text[i : i + 2] for i in range(0, len(text), 2)]


print("Example:")
print(list(split_pairs("abcd")))

assert list(split_pairs("abcd")) == ["ab", "cd"]
assert list(split_pairs("abc")) == ["ab", "c_"]
assert list(split_pairs("abcdf")) == ["ab", "cd", "f_"]
assert list(split_pairs("a")) == ["a_"]
assert list(split_pairs("")) == []

print("The mission is done! Click 'Check Solution' to earn rewards!")
